Closes #1636 add keep_vars to derive_locf_records

kaz462 opened the pull request:
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

bms63 commented:
@gg106046 what do you think?

bms63 commented:
Hi @kaz462 can you work through the check list, e.g. please add to the news, add a unit test, etc.  @gg106046 is super pro and can assist if you have any questions :)

kaz462 commented:
Thanks, @bms63 for checking! I've worked through the checklist and added news/unit test.
@gg106046 @bundfussr Currently, I added `keep_vars` to this function, and set the default as `NULL`. 
Another option I thought of is:
Incorporate `AVAL` into this argument, and rename  `keep_vars`  to something like `locf_vars` (set the default as `AVAL`), which might be more flexible, but not sure if it's necessary. what do you think?

bundfussr commented:
I would not add `AVAL` to `keep_vars` because it is handled differently. If `AVAL` is missing for a certain observation, an LOCF observation is created. If we add `AVAL` to `keep_vars`, it is not clear which variable should be used for checking on missing values. And what would be done, if `AVAL` is not specified for `keep_vars`?

Maybe we should add an `analysis_var` parameter and default it to `AVAL`. Then it would be possible to use `AVALC` instead of `AVAL`. What do you think?

kaz462 commented:
I agree, add `analysis_var` makes sense

bms63 commented:
Hey @kaz462 can you fix these merge conflicts when you have a chance??   The big one is that `vars()` has been replaced with `exprs()`

kaz462 commented:
Thanks for letting me know @bms63 
I've resolved the merge conflicts, but got the below error after getting the most recent updates from `devel` branch. could you help check?
```r
> devtools::load_all()
ℹ Loading admiral
Error in assert_filter_cond(enexpr(filter), optional = TRUE) : 
  is_quosure(arg) is not TRUE
In addition: Warning message:
replacing previous import ‘admiraldev::as_name’ by ‘rlang::as_name’ when loading ‘admiral’
```

bms63 commented:
@bundfussr are you happy with the updates.

bundfussr commented on R/derive_locf_records.R:
```diff
@@ -53,24 +64,24 @@
-#'   "CDISC01", "01-701-1028", "DIABP",     NA,        6, "WEEK 6",
-#'   "CDISC01", "01-701-1028", "SYSBP",    130,        0, "BASELINE",
-#'   "CDISC01", "01-701-1028", "SYSBP",    132,        2, "WEEK 2"
+#'   ~STUDYID, ~USUBJID, ~PARAMCD, ~PARAMN, ~AVAL, ~AVISITN, ~AVISIT,
```
Please preserve the alignment.

bundfussr commented on R/derive_locf_records.R:
```diff
@@ -93,25 +104,39 @@
+      chr2vars(colnames(dataset_expected_obs))
+    )
   )
+  assert_data_frame(
```
We do not need a second call. You can add `keep_vars` to the call above.

bundfussr commented on R/derive_locf_records.R:
```diff
@@ -17,24 +17,35 @@
-#'   in the input dataset or for which `AVAL` is NA for the corresponding observation
-#'   in the input dataset. Only variables specified in `by_vars` will be populated
-#'   in the newly created records.
+#'   in the input dataset or for which `analysis_var` is NA for the corresponding observation
```
Please use backticks for `NA`.

bundfussr commented on R/derive_locf_records.R:
```diff
@@ -93,25 +104,39 @@
     dataset,
-    required_vars = expr_c(by_vars, extract_vars(order), chr2vars(colnames(dataset_expected_obs)))
+    required_vars = expr_c(
+      by_vars, analysis_var, order,
```
Please use `extract_vars(order)` instead of `order`.  Otherwise the code fails if order is something like `desc(ASEQ)`.

kaz462 commented:
thanks for the review and comments! I addressed the above comments, and also updated below two
1. removed prefix `tibble::`
2. updated the description of input parameter `dataset`
